### ISSUE TITLE
pagination for comments

### ISSUE THREAD

**Issue by huumn** [stackernews/stacker.news#1470]
The time has come. We should probably limit comments to displaying 100 initially.

**Comment by ekzyis** [comment]
> What's the correct way to modify this function including options to paginate the results or, more in general, to declare or modify this kind of parametric queries?

We use Prisma for database migrations, see the [corresponding section in the README](https://github.com/stackernews/stacker.news?tab=readme-ov-file#database-migrations). Does that help?

**Comment by huumn** [comment]
Comments are queried in a recursive stored procedure that uses a temporary table. 

Completing this issue will require a migration that redefines the stored procedure.

**Comment by aegroto** [comment]
Hello, sorry for the late reply but I have been busy with a job and didn't have a chance to look further into this issue. I am exploring possibilities to fetch comments right now. Is there any specific reason you were not able to use prisma's find methods such as ```findMany``` and had to call raw queries?

**Comment by huumn** [comment]
Querying a relational tree and simultaneous aggregates is beyond what prisma is capable of.

**Comment by aegroto** [comment]
Sorry if I comment here again but I am still a novice with Prisma and PostgreSQL. I have written this function as a test:

```
CREATE OR REPLACE FUNCTION paginated_item_comments(item_id INT)
  RETURNS table ("text" text, "path" text)
  LANGUAGE sql IMMUTABLE STRICT PARALLEL SAFE
BEGIN ATOMIC
    SELECT "text", "path"
    FROM "Item"
    WHERE CAST("Item"."path" as text) LIKE item_id || '.%';
END;
```

This is doing part of the job (no pagination not specific handling for comments hierarchy yet, I was just playing around). I've tested it by manually registering it with ```sndev psql```, but how could it be included in a migration by Prisma? Sorry for the stupid question, but I haven't been able to figure out how the ```prisma/schema.prisma``` defines which functions are available in the database.

**Comment by huumn** [comment]
I don't have the bandwidth to teach you new tools. We have many issues that are easier than this one. I recommend doing those instead.

**Comment by aegroto** [comment]
I can understand your point of view, I didn't mean to ask you for assistance on the tools themselves but on the structure of the project instead, and I am sorry for the misunderstanding. I suggest you to provide some more documentation on how the database schema has been designed, if you have any spare time in the future, as I found understanding this aspect blocking and other willing contributors may to.

**Comment by ekzyis** [comment]
> I suggest you to provide some more documentation on how the database schema has been designed, if you have any spare time in the future, as I found understanding this aspect blocking and other willing contributors may to.

Is the section about [database migrations](https://github.com/stackernews/stacker.news?tab=readme-ov-file#database-migrations) in the README and the link to the Prisma docs not enough?

**Comment by aegroto** [comment]
@ekzyis Thanks for your interest. I will spend some time studying Prisma and be back on this issue if I will feel more confident with it.

**Comment by huumn** [comment]
We have some items with more than 100 items.

**Comment by cyphercosmo** [comment]
[Found them](https://stacker.news/top/posts/forever?by=comments)

**Comment by cyphercosmo** [comment]
Before I proceed, here's what I'm thinking: I need to change the producedure [`item_comments_zaprank_with_me`] adding pagination and limit. The complications I'm anticipating is the fact that the function is recursive and aggregates the final result meant to be returned, so I might have to be a little creative like wrapping this function and having the outer one do the filtering since we can't rely on the db itself to it for us.

> A side note: I tried seeding my database but the command `DATABASE_URL=postgresql://sn:password@localhost:5431/stackernews?schema=public node prisma/seed.js` after replacing the connection string with a working one completes the execution with an error. I did observe some records in the database, the table item is full of records but pretty much all the other tables are empty.

Obviously once that function is ready to accept the pagination arguments we will need to adjust the "wiring" all the way down to the frontend where we issue the GraphQL query. My idea is to follow the same pattern of having a button at the bottom to load more, the frontend will then maintain the state of which page it is and continues to query the next page as people press the button.

I can't really think of an attack vector introduced by this change, people could play with the arguments and send some ridiculous numbers which could break the slicing part of the algorithm in the end, for that we can do some input sanitization and assume certain defaults / limits.

Last but not least, given the tree nature of the "comments" which are just nested items, we can take different tree traversal approaches to determine where to stop because a single branch might go over or stay under the limit and I image we wouldn't want to show a branch partially and then as people click the button show the rest of the branch, probably staying under but as close as possible as we show entire branches.

Thoughts? Sorry for the long message.

**Comment by huumn** [comment]
It's hard to vet all this text so I won't, but based on what I skimmed, it sounds like you roughly have the obvious parts right.

> A side note: I tried seeding my database but the command DATABASE_URL=postgresql://sn:password@localhost:5431/stackernews?schema=public node prisma/seed.js after replacing the connection string with a working one completes the execution with an error. I did observe some records in the database, the table item is full of records but pretty much all the other tables are empty.

The docker environment provides a seed already. That seed is for use outside of the docker environment and is more or less deprecated.

**Comment by ekzyis** [comment]
> so I might have to be a little creative like wrapping this function and having the outer one do the filtering since we can't rely on the db itself to do it for us.

This sounds like this approach won't fix the performance problem though since after the filtering is applied, the expensive query already ran. Not 100% sure though, I am assuming the performance problem is caused in `item_comments_zaprank_with_me`.

**Comment by cyphercosmo** [comment]
One last thing @huumn, I created the below representation so you can visualize what I'm saying and agree with my suggestion.

![item tree](https://github.com/user-attachments/assets/d425783a-228a-457c-a70e-a760256e4498)

Now imagine that one of those items with 100+ comments could look something like that or even a lot more complex (like [this](https://stacker.news/items/661571)), so when we are returning the pages my suggestion would be to always return full clusters if adding a new cluster would go above the offset, we just don't want to show an incomplete cluster.

Do you agree?

**Comment by cyphercosmo** [comment]
> This sounds like this approach won't fix the performance problem though since the filtering is applied after the expensive query already ran. We want pagination because pages with a lot of comments load very slow.

Yes, I realized that, that's why I'll need to be creative and maybe pass some state like we pass the level to help the recursion know when to stop if we go over the limit. The skipping the first x records if we were on page y is something that I'll have to think about but I'm sure I'll find a way.

> Not 100% sure though, my objection assumes the performance problem is in `item_comments_zaprank_with_me`.

I love the fact you explicitly state your assumption, I'm thinking I need to mimic the same issue locally so I'll create an item with thousands of "comments" and use that to validate my changes.

I've been thinking of opening an issue to suggest some code improvements, like introducing tests to act as feedback loops for developers to know whether their change broke something or not.

**Comment by huumn** [comment]
> I've been thinking of opening an issue to suggest some code improvements, like introducing tests to act as feedback loops for developers to know whether their change broke something or not.

We have discussed this many times before. Please don't create an issue for it.

**Comment by huumn** [comment]
It might be best to start with an easier issue first if this one doesn't seem straightforward to you conceptually.

**Comment by cyphercosmo** [comment]
I believe I fully understand it, it's pretty easy actually.. but I was asking questions to align expectations. The issue spec wasn't very detailed.

**Comment by huumn** [comment]
> The issue spec wasn't very detailed.

Fair.
